### PR TITLE
fix(s3-deployment): fix passing system metadata to deployment

### PR DESCRIPTION
**Pull request by AlexCheema**
Fixes https://github.com/aws/aws-cdk/issues/4665
----
The AWS CLI list of commands generated when using system-defined metadata bundled together pairs of arguments, which caused the CLI to give e.g. `Unknown options: --cache-control 'max-age=600'` as an error. The fix is to ensure that each argument is a separate string in the list of AWS CLI commands.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by AlexCheema** [comment]
I'm unable to run the integration tests as I get this error when running `build.sh`: `@aws-cdk/cx-api: [2019-10-26T16:03:30.441] [ERROR] jsii/compiler - Error during type model analysis: TypeError: Cannot read property 'kind' of undefined`

File: packages/@aws-cdk/aws-s3-deployment/test/integ.bucket-deployment.ts [hunk 1]
```diff
@@ -27,6 +28,17 @@ class TestBucketDeployment extends cdk.Stack {
       destinationKeyPrefix: 'deploy/here/',
       retainOnDelete: false, // default is true, which will block the integration test cleanup
     });
+
+    const bucket3 = new s3.Bucket(this, 'Destination3');
```
Comment (eladb) [inline packages/@aws-cdk/aws-s3-deployment/test/integ.bucket-deployment.ts]: I am expecting to see a change in `.expected.json`

Comment (AlexCheema) [reply packages/@aws-cdk/aws-s3-deployment/test/integ.bucket-deployment.ts]: Please see previous comment. I'm unable to run integration tests. Someone else will have to jump in and update the `.expected.json`.

**Comment by AlexCheema** [comment]
> A few administrative notes (see PR checklist):
> 
> * The PR description should include what was the root cause of the bug and how it was addressed
> * The module name in the PR title should not have "aws-" prefix
> 
> Also, please also include a unit test that verifies this. Integration tests are not as powerful to protect against regressions.

Updated the PR description.

This bug specifically showed up when running actual AWS CLI commands.

**Comment by eladb** [comment]
More test updates?

**Comment by AlexCheema** [comment]
> More test updates?

It seems to me that it would convolute the tests, testing for this specific case.
The source of the bug was that I included pairs of AWS commands in each entry in the list of commands. Since we use this line `subprocess.check_call(["python3", aws] + list(args))` to invoke the cli, the pair of arguments were treated as a single argument.
Currently the tests do not preserve the lists when using `assertAwsCommands` - it just checks for matching space-separated strings.
The fix would be to instead of checking a string for each AWS command, preserve the list of AWS commands and check for list-equality.
Let me know if you think it's worth changing.

**Comment by AlexCheema** [comment]
@eladb just bumping this. I'm happy to implement the changes mentioned in my previous comment, I just want to check if you think this is a good direction.

**Comment by eladb** [comment]
What's the status here?

**Comment by AlexCheema** [comment]
> What's the status here?

Please read my previous comments. I'm awaiting your response for how to proceed.

**Comment by eladb** [comment]
> The fix would be to instead of checking a string for each AWS command, preserve the list of AWS commands and check for list-equality.

I think that's the right thing to do. It will make the tests more robust

**Comment by eladb** [comment]
@AlexCheema any updates here?

**Comment by eladb** [comment]
closing as stale, reopen as needed

**Pull request closed without merging**